### PR TITLE
Add svg - and remove unclear parts.

### LINKED ISSUE #174: Add support for SVG and MathML
**Reported by dietmarw on 27 May 2009 07:47 UTC**
Although probably being a tool issue, this ticket should act as an reminder to add SVG and MathML support in some future version of Modelica

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/174

### PR DESCRIPTION
**Pull request by HansOlsson**
Adding SVG as decided a long time ago, and removing unclear text.
Closes #174

### CODE REVIEW

File: chapters/annotations.tex [hunk 1]
```diff
@@ -887,8 +887,8 @@ \subsubsection{Bitmap}\doublelabel{bitmap}
 When the attribute \lstinline!fileName! is specified, the string refers to an
 external file containing image data. The mapping from the string to the
 file is specified for some URIs in \autoref{external-resources}. The supported file
-formats include \lstinline!PNG!, \lstinline!BMP! and \lstinline!JPEG!, other supported file formats are
-unspecified.
+formats include \lstinline!PNG!, \lstinline!BMP!, \lstinline!JPEG!, 
+and \lstinline!SVG! (provided that the outermost svg-element specify the size, i.e. width and height).
```
Comment (dietmarw) [inline chapters/annotations.tex]:
```suggestion
and \lstinline!SVG! (provided that the outermost svg-element specify the size, i.e., width and height).
```

File: chapters/annotations.tex [hunk 2]
```diff
@@ -887,8 +887,8 @@ \subsubsection{Bitmap}\doublelabel{bitmap}
 When the attribute \lstinline!fileName! is specified, the string refers to an
 external file containing image data. The mapping from the string to the
 file is specified for some URIs in \autoref{external-resources}. The supported file
-formats include \lstinline!PNG!, \lstinline!BMP! and \lstinline!JPEG!, other supported file formats are
-unspecified.
+formats include \lstinline!PNG!, \lstinline!BMP!, \lstinline!JPEG!,
+and \lstinline!SVG! (provided that the outermost svg-element specify the size, i.e., width and height).
```
Comment (beutlich) [inline chapters/annotations.tex]:
```suggestion
and \lstinline!SVG! (provided that the outermost svg-element specifies the size, i.e., width and height).
```

File: chapters/annotations.tex (hunk 1 above)
Comment (henrikt-ma) [inline chapters/annotations.tex]: Given the complexity of the SVG format, I think it would be good to be very precise regarding how the size is specified:
> provided that the outermost \lstinline!svg! element has a size determined by its \lstinline!width! and \lstinline!height! attributes

Since absence of `width` and `height` means the same as 100%, I would think that our intention here is also to require that `width` and `height` are set to absolute lengths.

Comment (henrikt-ma) [reply chapters/annotations.tex]: Also, what about unitless lengths?  If they should be allowed, I think we need to specify what the _current user coordinate system_ is in this case.

Comment (sjoelund) [reply chapters/annotations.tex]: We only need to know the ratio of width/height for our purposes

Comment (HansOlsson) [reply chapters/annotations.tex]: Just to be clear width and height of "outermost svg element" is part of the SVG-specification:
http://www.w3.org/TR/SVG11/struct.html#SVGElementHeightAttribute

> height = "<length>"
> For outermost svg elements, the intrinsic height of the SVG document fragment

However, it _seems_ that it can be in percentages (not sure if viewBox is needed for this or not); so I added "absolute size" to clarify that.

Comment (henrikt-ma) [reply chapters/annotations.tex]: Sure, I just wanted it to be explicit that we are taking precisely about these attributes, and not just _width_ and _height_ in general.

Comment (HansOlsson) [reply chapters/annotations.tex]: Changed to lstinline for width and height.

**Comment by HansOlsson** [comment]
A bit of discussion about size, possibilities are:
Don't add svg: 1
Add with width/height and possibly later revisit (to generalize width/height): 1
Figure out how to handle width/height before adding: 8
Henrik volunteers.

**Comment by henrikt-ma** [comment]
I have now convinced myself that we shouldn't require neither `width`, `height`, nor `viewBox`.  I have made example SVG images and an HTML page with various `` tags for showing the images, and I'm happy to share them with anyone how is interested, or show the results in a screen sharing session.

Anyway, the main conclusions are these:
- As soon as there is a `viewBox`, this part of the SVG canvas is always mapped to the `extent` of the `BitMap`, and `width` and `height` are ignored for rendering purposes.
- The `width` and `height` only impact the cropping of the SVG canvas when there is no `viewBox` attribute, and this is then what gets mapped to the `extent` of the `BitMap`.
- Even the case of having neither of `width`, `height`, nor `viewBox` is well defined.  Here, it is the `extent` of the `BitMap` that determines the cropping of the unscaled SVG canvas.  This may not be a very useful case, but there is no technical reason for not allowing it.
- The `width` and `height` are always meaningful for setting (and presenting) the `extent` of the `BitMap` in a fixed relation to the intrinsic size of the SVG image, but doing this can be left as a quality of GUI implementation thing.  (Later, we could also add language support for this by allowing an `relativeExtent` as an alternative to `extent`, where the `relativeExtent` would give the upper left corner of the extent, together with two scaling factors for converting the intrinsic width and height to the width and height of the `BitMap`.)
- The SVG attribute `preserveAspectRatio` impacts the mapping of the cropped part of the SVG canvas to the `extent` of the `BitMap`, but there is no need to mention it in the `BitMap` specification.  However, when `preserveAspectRatio` is set to `meet`, a GUI could assist the user by only allowing the `extent` to be set with the same aspect ratio as the intrinsic size of the SVG image.

In other words, the `BitMap` specification should just list SVG as a supported format without restrictions.

**Comment by henrikt-ma** [comment]
By the way, if we are serious about the support for SVG in `BitMap`, I think it would be very helpful to have a package somewhere with test models that illustrate different ways of combining `width`, `height`, `viewBox` and `preserveAspectRatio`.  (Of course, the package should somehow also include screen shots with the expected results for all test cases.)

**Comment by dietmarw** [comment]
> I'm happy to share them with anyone how is interested, or show the results in a screen sharing session.

@henrikt-ma I'd be interested.

**Comment by HansOlsson** [comment]
> 
> 
> By the way, if we are serious about the support for SVG in `BitMap`, I think it would be very helpful to have a package somewhere with test models that illustrate different ways of combining `width`, `height`, `viewBox` and `preserveAspectRatio`. (Of course, the package should somehow also include screen shots with the expected results for all test cases.)

A good idea, and I would like to be able to test that.

**Comment by svorkoetter** [comment]
> A good idea, and I would like to be able to test that.

Seconded.

**Comment by henrikt-ma** [comment]
> > I'm happy to share them with anyone how is interested, or show the results in a screen sharing session.
> 
> @henrikt-ma I'd be interested.

When would be a good time for a screen sharing session?  Today, I'm only available until lunch, but I have no meetings planned either Monday or Tuesday next week.

**Comment by henrikt-ma** [comment]
OK, let's do it now:
https://wolfram.zoom.us/j/511560589
